src/main/java/net/minecraftforge/common/tags/ForgeTagBuilder.java

```diff
@@ -8,45 +12,123 @@
 import net.minecraft.tags.Tag.ITagEntry;
 import net.minecraft.util.JsonUtils;
 import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.registries.IForgeRegistryEntry;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 import javax.annotation.ParametersAreNonnullByDefault;
-import java.util.Collection;
-import java.util.Comparator;
-import java.util.LinkedHashSet;
-import java.util.Set;
+import java.util.*;
+import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Predicate;
+
 @MethodsReturnNonnullByDefault
 @ParametersAreNonnullByDefault
-public final class ForgeTagBuilder<T> extends Tag.Builder<T> {
-    private static final Logger LOGGER = LogManager.getLogger();
-    private boolean resolved;
-    private final TagFactory<T> factory;
+public final class ForgeTagBuilder<T> extends Tag.Builder<T>
+{
+    public static Comparator<IForgeRegistryEntry<?>> registryNameComparator()
+    {
+        return Comparator.nullsFirst(Comparator.<IForgeRegistryEntry<?>, ResourceLocation>comparing(IForgeRegistryEntry::getRegistryName, Comparator.nullsFirst(Comparator.naturalOrder())));
+    }
 
-    public static <T> ForgeTagBuilder<T> vanillaTagBuilder() {
-        return new ForgeTagBuilder<>(new TagFactory<T>() {
-            @Override
-            public Tag<T> build(ResourceLocation location, Set<ITagEntry<T>> iTagEntries, boolean preserveOrder) {
-                return new Tag<>(location,iTagEntries,preserveOrder);
-            }
+    public static <T> Comparator<Tag<T>> tagIdComparator()
+    {
+        return Comparator.nullsFirst(Comparator.<Tag<T>, ResourceLocation>comparing(Tag::getId, Comparator.nullsFirst(Comparator.naturalOrder())));
```

Comment:
Why do we need null first? If null happens, we should just crash.

Reply:
Ok